### PR TITLE
Add comments expire date

### PR DESCRIPTION
**Pull request by vitormattos**
Related: https://github.com/nextcloud/spreed/pull/7327

### CODE REVIEW

File: lib/private/Comments/Manager.php [hunk 1]
```diff
@@ -1642,4 +1646,21 @@ public function load(): void {
 		$this->initialStateService->provideInitialState('comments', 'max-message-length', IComment::MAX_MESSAGE_LENGTH);
 		Util::addScript('comments', 'comments-app');
 	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function deleteMessageExpiredAtObject(string $objectType, string $objectId): int {
```
Comment (nickvergessen) [inline lib/private/Comments/Manager.php:1653]: please move up to the other delete functions (then you also might see the thing about returning a boolean which I commented on the interface)

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/public/Comments/ICommentsManager.php [hunk 2]
```diff
@@ -482,4 +482,15 @@ public function resolveDisplayName($type, $id);
...
 	 */
 	public function load(): void;
+
+	/**
+	 * Delete comments with field expire_date less than current date
+	 * Only will delete the message related with the object.
+	 *
+	 * @param string $objectType the object type (e.g. 'files')
+	 * @param string $objectId e.g. the file id
+	 * @return boolean true if rows was deleted
```
Comment (nickvergessen) [inline lib/public/Comments/ICommentsManager.php:492]:
```suggestion
	 * @return boolean true if at least one row was deleted
```
or
```suggestion
	 * @return boolean true if rows were deleted
```

**Comment by nickvergessen** [comment]
Tested with our instance:
```sql
SELECT COUNT(*) FROM oc_comments;
+----------+
| COUNT(*) |
+----------+
|  1059368 |
+----------+

ALTER TABLE `oc_comments` ADD `expire_date` DATETIME NULL AFTER `reactions`; 
Query OK, 0 rows affected (0.028 sec)
Records: 0  Duplicates: 0  Warnings: 0

ALTER TABLE `oc_comments` ADD INDEX `expire_date` (`expire_date`); 
Query OK, 0 rows affected (2.084 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

So sounds good.

**Comment by vitormattos** [comment]
Now, can I squash the commits?

**Comment by nickvergessen** [comment]
> Now, can I squash the commits?

Sounds good to me

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
you need to run `bash build/autoload...` and commit the changed files in `lib/composer/composer/autoload_*` (not the `installed.php` and `InstalledVersion.php`)

File: tests/lib/Comments/ManagerTest.php [hunk 3]
```diff
@@ -701,6 +703,51 @@ public function testDeleteCommentsAtObject() {
...
+			$this->connection,
+			$this->createMock(LoggerInterface::class),
+			$this->createMock(IConfig::class),
+			Server::get(ITimeFactory::class),
+			new EmojiHelper($this->connection),
+			$this->createMock(IInitialStateService::class)
+		);
+
+		// just to make sure they are really set, with correct actor data
+		$comment = $manager->get(strval($ids[1]));
```
Comment (blizzz) [inline tests/lib/Comments/ManagerTest.php:725]:
```suggestion
		$comment = $manager->get((string)$ids[1]);
```

micro optimization :)

File: tests/lib/Comments/ManagerTest.php [hunk 4]
```diff
@@ -701,6 +703,51 @@ public function testDeleteCommentsAtObject() {
...
+		$this->assertSame($comment->getObjectId(), 'file64');
+
+		$deleted = $manager->deleteMessageExpiredAtObject('files', 'file64');
+		$this->assertTrue($deleted);
+
+		$deleted = 0;
+		$exists = 0;
+		foreach ($ids as $id) {
+			try {
+				$manager->get(strval($id));
```
Comment (blizzz) [inline tests/lib/Comments/ManagerTest.php:736]:
```suggestion
				$manager->get((string)$id);
```

**Review by blizzz — CHANGES_REQUESTED** [review]
CI is unhappy

expire_date is also not exposed in DAV API, making it inconsistent with the other features. Cf. apps/dav/lib/Comments/CommentNode.php

**Comment by nickvergessen** [comment]
> expire_date is also not exposed in DAV API

Neither is the optional `refernce_id` that we added for talk specifically. I think that's fine.

**Comment by blizzz** [comment]
> > expire_date is also not exposed in DAV API
> 
> Neither is the optional `refernce_id` that we added for talk specifically. I think that's fine.

Does not mean it is not worth to add it there as well. The reference id has of course less meaning for file comments. The expiry date would be more relevant, but on the other hand the web ui does not offer this option either. Thus, I also agree it is okay to leave it out, nevertheless worth to point out that this is done deliberately.

File: tests/lib/Comments/ManagerTest.php (hunk 3 above)
Comment (vitormattos) [reply tests/lib/Comments/ManagerTest.php]: I did other test to check the opcode in site 3v4l.org using the follow code:

```php
echo (string) 1;
echo strval(1);
```
...and the opcode of second row have one more step, a return:

```
line      #* E I O op                           fetch          ext  return  operands
-------------------------------------------------------------------------------------
    3     0  E >   CAST                                          6  ~0      1
          1        ECHO                                                     ~0
    4     2        CAST                                          6  ~1      1
          3        ECHO                                                     ~1
          4      > RETURN                                                   1
```

https://3v4l.org/ZGBcX/vld

Comment (blizzz) [reply tests/lib/Comments/ManagerTest.php]: Ok, it was intval vs int-casting. Never mind then!

**Review by blizzz — APPROVED** [review]

**Comment by nickvergessen** [comment]
Missing a version bump

**Pull request merged**